modules/3d/include/opencv2/3d.hpp

```diff
@@ -468,77 +468,308 @@ can be found in:
+This class does not support fixed/masked variables, this should also be implemented in child classes.
+
+A Levenberg-Marquadt algorithm locally minimizes an objective function value (aka energy, cost or error) starting from
+current param vector.
+To do that, at each iteration it repeatedly calculates the energy at probe points until it's reduced.
+To calculate a probe point, a linear equation is solved: (J^T*J + lambda*D)*dx = -J^T*b where J is a function jacobian,
+b is a vector of residuals (aka errors or energy terms), D is a diagonal matrix generated from J^T*J diagonal
+and lambda changes for each probe point. Then the resulting dx is "added" to current variable and it forms
+a probe value. "Added" is quoted because in some groups (e.g. SO(3) group) such an increment can be a non-trivial operation.
+
+For more details, please refer to Wikipedia page (https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm).
 */
-class CV_EXPORTS LMSolver : public Algorithm
+class CV_EXPORTS LevMarqBase
 {
 public:
-    class CV_EXPORTS Callback
+
+    /** @brief Optimization report
+
+    The structure is returned when optimization is over.
+    */
+    struct CV_EXPORTS Report
+    {
+        Report(bool isFound, int nIters, double finalEnergy) :
+            found(isFound), iters(nIters), energy(finalEnergy)
+        { }
+        // true if the cost function converged to a local minimum which is checked by check* fields, thresholds and other options
+        // false if the cost function failed to converge because of error, amount of iterations exhausted or lambda explosion
+        bool found;
+        // amount of iterations elapsed until the optimization stopped
+        int iters;
+        // energy value reached by the optimization
+        double energy;
+    };
+
+
+    /** @brief Structure to keep LevMarq settings
+
+    The structure allows a user to pass algorithm parameters along with their names like this:
+    @code
+    MySolver solver(nVars, callback, MySolver::Settings().geodesicS(true).geoScale(1.0));
+    @endcode
+
+    */
+    struct CV_EXPORTS Settings
+    {
+        Settings() :
+            jacobiScaling(false),
+            upDouble(true),
+            useStepQuality(true),
+            clampDiagonal(true),
+            stepNormInf(false),
+            checkRelEnergyChange(true),
+            checkMinGradient(true),
+            checkStepNorm(true),
+            geodesic(false),
+            hGeo(1e-4),
+            geoScale(0.5),
+            stepNormTolerance(1e-6),
+            relEnergyDeltaTolerance(1e-6),
+            minGradientTolerance(1e-6),
+            smallEnergyTolerance(0), // not used by default
+            maxIterations(500),
+            initialLambda(0.0001),
+            initialLmUpFactor(2.0),
+            initialLmDownFactor(3.0)
+        { }
+
+        bool operator==(const Settings& other) const
```

Comment:
If users don't need this functionality, then it is better to create internal function.

Reply:
Removed
(was made to compare a passed arg with default settings constant, default arg is used instead)